tests/simulator_test.cpp

```diff
@@ -2,82 +2,284 @@
+const Vec3 kVecY(0, 1, 0);
+
+float deg2rad(float deg) { return 2 * kPi * deg / 360; }
+
+// TODO: add collision point check for every ray hit
+
+class FakeModel : public ModelInterface {
+
+public:
+  explicit FakeModel(bool empty) : empty_(empty){};
+  std::vector<objects::TriangleObj *> triangles() const override { return {}; }
+  float height() const override { return 0; }
+  float sideSize() const override { return 0; }
+  bool empty() const override { return empty_; }
+
+private:
+  bool empty_;
+};
+
+class EnergyCollectorTest : public ::testing::Test {
+protected:
+  // performs ray hit at at given energy collectors, |hitData| is modified to
+  // hold information where ray hit energyCollector. Returns true if hit
+  // occurred, false when there was no hit
+  [[nodiscard]] bool performHitCollector(
+      const std::vector<std::unique_ptr<objects::EnergyCollector>>
+          &energyCollectors,
+      const Ray &ray, RayHitData *hitData) {
+
+    for (const auto &collector : energyCollectors) {
+      if (collector->hitObject(ray, kSkipFrequency, hitData)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  void
+  printCollectors(const std::vector<std::unique_ptr<objects::EnergyCollector>>
+                      &energyCollectors) const {
+    for (const auto &collector : energyCollectors) {
+      std::cout << "Origin: " << collector->getOrigin() << "\n"
+                << "Radius: " << collector->getRadius() << "\n";
+    }
+  }
+
+  std::pair<objects::EnergyCollector, objects::EnergyCollector>
+  topCollectorsXAxis(
+      const std::vector<std::unique_ptr<objects::EnergyCollector>>
+          &energyCollectors) {
+    auto outputCollectors =
+        std::make_pair(objects::EnergyCollector(Vec3(0, 0, 0), 1),
+                       objects::EnergyCollector(Vec3(0, 0, 0), 1));
+    for (const auto &collector : energyCollectors) {
+      if (collector->getOrigin().z() >=
+              outputCollectors.first.getOrigin().z() &&
+          collector->getOrigin().x() ==
+              outputCollectors.first.getOrigin().x()) {
+        outputCollectors.second = outputCollectors.first;
+        outputCollectors.first = *(collector.get());
+      }
+    }
+    return outputCollectors;
   }
+};
+TEST_F(EnergyCollectorTest, ThrowExceptionWhenInvalidNumCollector) {
 
-  ASSERT_EQ(manager.getEnergyCollectors().size(), kPopulation);
+  const FakeModel nonEmptyModel(false);
 
+  const int invalidNumCollectors = 38;
+  ASSERT_EXCEPTION(
+      buildCollectors(nonEmptyModel, invalidNumCollectors),
+      std::invalid_argument,
+      "numCollectors or numCollectors-1 has to be divisible by 4, got "
+      "numCollectors = 38");
+
+  const int numCollectorLessThenOne = 3;
+  ASSERT_EXCEPTION(buildCollectors(nonEmptyModel, numCollectorLessThenOne),
+                   std::invalid_argument, "numCollectors: 3 is less then 4");
+
+  // Test case when numCollector - 1 % 4 = 0
+  const int validNumCollectorCase1 = 37;
```

Comment:
You can just inline the value. If this is valid then we see this with EXPECT_NO_THROW. With inlined value we already know what the param is when this fails, but before you would have togo to the code.

Reply:
I'm not sure if I understand this correctly. I've never made any inline member. So if I understand correclty, you wan me to have case1 and case2 have the same name, and with inline keyword let compiler decide which value should he use? 

Can you explain me a little bit more about this? ;-)